scripts/release.py

```diff
@@ -0,0 +1,276 @@
+        if rel_type == "minor":
+            # leave only minor releases
+            tags = [(entry, ver) for entry, ver in tags if ver.micro == 0]
+        else:
+            assert rel_type == "patch"
+        prev_ref, prev_ver = max(tags, key=lambda pair: pair[1])
+        return prev_ref, self.repo.references[prev_ref].peel(), prev_ver
+
+    def get_commits_upto(self, stop_commit):
+        history = []
+        for obj in self.repo.walk(self.repo.head.target):
+            if obj.id == stop_commit.id:
+                break
+            history.append(obj)
+        else:
+            raise ValueError("Current HEAD is not derived from previous release")
+        return history
+
+    def ensure_title_link(self, obj: pygit2.Commit):
+        title = obj.message.splitlines()[0]
+        if not re.match(r".*\(#(\d+)\)$", title):
+            title += f" ({obj.short_id})"
+        return title
+
+
+def make_notes(args):
+    wrapper = GitWrapper()
+    release_type = "minor" if wrapper.is_on_master() else "patch"
+    sys.stderr.write(f"Detected release type: {release_type}\n")
+
+    prev_ref, prev_commit, prev_ver = wrapper.get_previous_release(release_type)
+    sys.stderr.write(f"Previous {release_type} release: {prev_ref}\n")
+
+    next_major, next_minor, next_patch = prev_ver.release
+    if release_type == "minor":
+        next_minor += 1
+    elif release_type == "patch":
+        next_patch += 1
+    else:
+        raise ValueError(f"Unexpected release type: {release_type}")
+    next_ver = version.Version(f"{next_major}.{next_minor}.{next_patch}")
+
+    sys.stderr.write(f"Computing release notes for {prev_ver} -> {next_ver}...\n")
+    try:
+        history = wrapper.get_commits_upto(prev_commit)
+    except ValueError as ex:
+        sys.stderr.write(
+            f"{ex}: did you forget to checkout correct branch or pull tags?"
+        )
+        return 1
+    if not history:
+        sys.stderr.write(f"No commits since {prev_ver} found, nothing to generate!\n")
+        return 1
+
+    titles = collections.defaultdict(list)
+    people = set()
+    email2commit, email2pr = {}, {}
+    for obj in history:
+        title = obj.message.splitlines()[0]
+        titles[title.split("-")[0]].append(obj)
+        new_people = set(
+            re.findall(
+                r"(?:(?:Signed-off-by|Co-authored-by):\s*)([\w\s,]+?)\s*<([^>]+)>",
+                obj.message,
+            )
+        )
+        for _, email in new_people:
+            email2pr[email] = obj.id
+        people |= new_people
+        email2commit[obj.author.email] = obj.id
+    sys.stderr.write(f"Found {len(history)} commit(s) since {prev_ref}\n")
+
+    sys.stderr.write("Resolving contributors...\n")
+    user_resolver = GithubUserResolver(email2commit, args.token)
+    logins, unknowns = user_resolver.resolve(people)
+    new_logins, unknowns = user_resolver.resolve_by_reviews(unknowns, email2pr)
+    logins |= new_logins
```

Comment:
Why are we resolving the Github usernames? Is this necessary?

Reply:
Because we list the contributors in our release notes by tagging them, and one tags by Github usernames.
See "Contributors" section [in 0.17.0 notes](https://github.com/modin-project/modin/releases/tag/0.17.0) as an example.

Reply:
Gotcha. Makes sense.